pkg/controller/kogitoapp/resource/manage_resources.go

```diff
@@ -0,0 +1,360 @@
+		if buildConfigS2I.Spec.Strategy.SourceStrategy.From.Name != s2iImageName {
+			log.Debug("Changes detected in 'ImageName'.", " OLD - ", buildConfigS2I.Spec.Strategy.SourceStrategy.From.Name, " NEW - ", s2iImageName)
+
+			buildConfigS2I.Spec.Strategy.SourceStrategy.From.Name = s2iImageName
+			updateS2I = true
+		}
+		if buildConfigS2I.Spec.Strategy.SourceStrategy.From.Namespace != s2iImageNamespace {
+			log.Debug("Changes detected in 'ImageNamespace'.", " OLD - ", buildConfigS2I.Spec.Strategy.SourceStrategy.From.Namespace, " NEW - ", s2iImageNamespace)
+
+			buildConfigS2I.Spec.Strategy.SourceStrategy.From.Namespace = s2iImageNamespace
+			updateS2I = true
+		}
+
+		envs := envToMap(instance.Spec.Build.Env)
+		if instance.Spec.Runtime == v1alpha1.QuarkusRuntimeType {
+			envs[nativeBuildEnvVarKey] = strconv.FormatBool(instance.Spec.Build.Native)
+		}
+		if !reflect.DeepEqual(envs, envVarToMap(buildConfigS2I.Spec.Strategy.SourceStrategy.Env)) {
+			log.Debug("Changes detected in 'Env'.", " OLD - ", buildConfigS2I.Spec.Strategy.SourceStrategy.Env, " NEW - ", envs)
+
+			buildConfigS2I.Spec.Strategy.SourceStrategy.Env = mapToEnv(envs)
+			updateS2I = true
+		}
+
+		if instance.Spec.Build.Incremental != *buildConfigS2I.Spec.Strategy.SourceStrategy.Incremental {
+			log.Debug("Changes detected in 'Incremental'.", " OLD - ", *buildConfigS2I.Spec.Strategy.SourceStrategy.Incremental, " NEW - ", instance.Spec.Build.Incremental)
+
+			buildConfigS2I.Spec.Strategy.SourceStrategy.Incremental = &instance.Spec.Build.Incremental
+			updateS2I = true
+		}
+
+		if instance.Spec.Build.GitSource.ContextDir != buildConfigS2I.Spec.Source.ContextDir {
+			log.Debug("Changes detected in 'ContextDir'.", " OLD - ", buildConfigS2I.Spec.Source.ContextDir, " NEW - ", instance.Spec.Build.GitSource.ContextDir)
+
+			buildConfigS2I.Spec.Source.ContextDir = instance.Spec.Build.GitSource.ContextDir
+			updateS2I = true
+		}
+
+		if *instance.Spec.Build.GitSource.URI != buildConfigS2I.Spec.Source.Git.URI {
+			log.Debug("Changes detected in 'GitSourceURI'.", " OLD - ", buildConfigS2I.Spec.Source.Git.URI, " NEW - ", *instance.Spec.Build.GitSource.URI)
+
+			buildConfigS2I.Spec.Source.Git.URI = *instance.Spec.Build.GitSource.URI
+			updateS2I = true
+		}
+		if instance.Spec.Build.GitSource.Reference != buildConfigS2I.Spec.Source.Git.Ref {
+			log.Debug("Changes detected in 'GitSourceRef'.", " OLD - ", buildConfigS2I.Spec.Source.Git.Ref, " NEW - ", instance.Spec.Build.GitSource.Reference)
+
+			buildConfigS2I.Spec.Source.Git.Ref = instance.Spec.Build.GitSource.Reference
+			updateS2I = true
+		}
+
+		if updateS2I {
+			log.Info("Updating BuildConfig for S2I")
+
+			if err := kubernetes.ResourceC(client).Update(buildConfigS2I); err != nil {
+				return updateS2I, err
+			}
+
+			if _, err := openshift.BuildConfigC(client).TriggerBuild(buildConfigS2I, instance.Name); err != nil {
+				return updateS2I, err
+			}
+
+			return updateS2I, nil
```

Comment:
In your tests, updating the BC instead of recreating it in case of updates was easier? I guess that we could safely delete the BC and recreate it in here since might have some sort of "unexpected" results if updated.

Reply:
@ricardozanini , I my test, after updating the BC, I explicitly [start/trigger a new Build](https://github.com/kiegroup/kogito-cloud-operator/blob/08df2ff172fb4f891bc02c9423c105bcc5851e3d/pkg/controller/kogitoapp/resource/manage_resources.go#L269) with the BuildConfig. Otherwise, the Build won't run. We can keep track of all the previous Builds if we don't recreate the BC.

Reply:
Seems fair having the build history. This is what an administrator would expect. Let’s leave this way then and keep an eye on every possible modification that we could do on BC and update accordingly. 

I’m afraid we need a step to check if the BC is meant to run native builds or not, right?

Reply:
@ricardozanini do you mean checking the BC after updating the CR to use native build? The [code](https://github.com/kiegroup/kogito-cloud-operator/blob/33e9918d0604a0593a609031c21facc071cb286f/pkg/controller/kogitoapp/resource/manage_resources.go#L227) checks if it is configured to use native build.

Reply:
Yeah, right. But the native flag build will use other image to build it.

For example: if I have a JVM build set on the BCS2I and I change my mind and change it to native, what happens?

The update hook should update `.Spec.Build.Native=true` and also performs the image build update. We could separate this use case into a function that is called during the BC creation and by the update hook. WDYT?

Reply:
This part is moved to the function [resolveRuntimeImage](https://github.com/kiegroup/kogito-cloud-operator/blob/33e9918d0604a0593a609031c21facc071cb286f/pkg/controller/kogitoapp/resource/build_config_runtime.go#L61) for the runtime BC creation. And the [update BC code](https://github.com/kiegroup/kogito-cloud-operator/blob/33e9918d0604a0593a609031c21facc071cb286f/pkg/controller/kogitoapp/resource/manage_resources.go#L295) calls this function to get the expected image, and compares this image with the image which is currently deployed. And also created the function [resolveS2IImage](https://github.com/kiegroup/kogito-cloud-operator/blob/aec718ab497fe4ab136f6a4a4faf69a57003c552/pkg/controller/kogitoapp/resource/build_config_s2i.go#L83) for the [s2i BC update](https://github.com/kiegroup/kogito-cloud-operator/blob/aec718ab497fe4ab136f6a4a4faf69a57003c552/pkg/controller/kogitoapp/resource/manage_resources.go#L203).